### PR TITLE
Select Component and CustomFieldInputSingleChoice Refactor

### PR DESCRIPTION
**Pull request by Bloodyaugust**
## Motivation
We want a single select for arbitrary data. It would be really great if we could leverage the work from CustomFieldInputSingleChoice. I know the single choice component has some specific code to deal with custom field logic, so if we were to leverage the component code, we need to figure out which parts belong in MDS (Generic) Select and which parts belong in MDS CustomFieldInputSingleChoice

## Acceptance Criteria 
- MDS Select
  - `name` prop and ref API (symmetry)
  - `value` prop and ref API (symmetry)
  - `onChange` prop API (called when value state changes)
  - `required` prop API
  - `children` prop API (as their options)

## PR upkeep checklist

- [x] Change log entry
- [x] Label(s)
- [x] Assignee(s)
- [x] Deployment URL: https://mavenlink.github.io/design-system/webinf-select-all-the-things-174890763
- [x] (When ready for review) Reviewer(s)
- [x] Green Bigmaven CI check
- [ ] DevQA on Bigmaven staging (e.g. does the work need to be imported in the internal design system?)

https://www.pivotaltracker.com/story/show/174890763

### CODE REVIEW

File: src/components/custom-field-input-single-choice/custom-field-input-single-choice.css [hunk 1]
```diff
@@ -1,28 +1,3 @@
-@import '../../styles/colors-v2.css';
-@import '../../styles/spacing.css';
-
 .container {
```
Comment (juanca) [inline src/components/custom-field-input-single-choice/custom-field-input-single-choice.css:1]: Do we still need this around for single choice? Looks like Select implements the same rule on its root node

File: src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx [hunk 2]
```diff
@@ -154,43 +72,21 @@ const CustomFieldInputSingleChoice = forwardRef(function CustomFieldInputSingleC
 
   return (
     <div ref={wrapperRef} className={props.className}>
```
Comment (juanca) [inline src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx:74]: Looks like we don't use `wrapperRef` anymore. We might even be able to throw away the className too.

File: src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx [hunk 3]
```diff
@@ -154,43 +72,21 @@ const CustomFieldInputSingleChoice = forwardRef(function CustomFieldInputSingleC
...
-            refs={refs}
-            value={value}
-          >
-            { listOptions(choices) }
-          </Listbox>
-        )
-      )}
+        value={value}
+      >
+        { listOptions(choices) }
```
Comment (juanca) [inline src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx:88]: Can we inline the choices generation or simplify the chain of functions into just 1 function call? i think we have a chain of functions that only really ever get called once -- an artifact of probably the previous large single choice component.

File: src/components/select/select.jsx [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+  useEffect(() => {
+    props.onChange({ target: selfRef.current });
+  }, [value]);
+
+  useEffect(() => {
+    updateOptionVisibility();
+  }, [searchValue, props.listOptionRefs]);
+
+  useEffect(() => {
```
Comment (juanca) [inline src/components/select/select.jsx:147]: Looking at this file, there is still a feeling of "woah, there's a lot going on here" -- and I think there isn't a way around it. But, I think an idea that might help:

Looks like we only use `updateOptionSelected` in this `useEffect`. We can just inline the function. Something like:

```suggestion
  useEffect(function updateOptionSelected() {
    ...
  }, [value, showOptions]);
```

Comment (juanca) [reply src/components/select/select.jsx:147]: If we like the idea, let's do the same for the other useEffects.

**Review by juanca — APPROVED** [review]
Overall, looks great. I think there are some refactors for the sake of readability / maintainability.

File: src/components/custom-field-input-single-choice/custom-field-input-single-choice.css (hunk 1 above)
Comment (Bloodyaugust) [reply src/components/custom-field-input-single-choice/custom-field-input-single-choice.css:1]: Good point, with that the whole style file can be removed.

File: src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx (hunk 2 above)
Comment (Bloodyaugust) [reply src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx:74]: I can see removing the `wrapperRef`, but I feel like we should maintain the `className` API to provide for style overrides in the future, and to keep parity with the other components.

File: src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx (hunk 3 above)
Comment (Bloodyaugust) [reply src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx:88]: Can do!

File: src/components/select/select.jsx (hunk 4 above)
Comment (Bloodyaugust) [reply src/components/select/select.jsx:147]: I'll try this fo sho

File: src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx [hunk 5]
```diff
@@ -216,7 +103,7 @@ CustomFieldInputSingleChoice.propTypes = {
 
 CustomFieldInputSingleChoice.defaultProps = {
   choices: [],
-  className: styles.container,
+  className: '',
```
Comment (juanca) [inline src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx:106]: Setting this to an empty string will override the default Select className. Is that intended?

Comment (Bloodyaugust) [reply src/components/custom-field-input-single-choice/custom-field-input-single-choice.jsx:106]: Noooope, definitely not! Fixing...

**Pull request merged**